File: backpac.js
/**
 * @param {number[]} values - آرایه‌ای از ارزش‌های اشیاء
 * @param {number[]} weights - آرایه‌ای از وزن‌های اشیاء
 * @param {number} capacity - ظرفیت کوله پشتی (حداکثر وزن مجاز)
 * @returns {number} - ارزش بیشینه‌ای که می‌توان در کوله پشتی قرار داد
 */

function knapsack(values,weights,capacity){
    const n =values.length; // count of items
    // 2D Array for save result
    const d2=Array.from({length:n+1},()=>Array(capacity+1).fill(0));

    // with Dynamic Algorithem push on Array
    for(let i=1;i<=n;i++){
        for(let w=1;w<=capacity;w++){
            if(weights[i-1]<=w){
                //if W(current weight) less or lower than i-1,calculate i-1 beetwen chose or not 
                d2[i][w]=Math.max(d2[i-1][w],d2[i-1][w-weights[i-1]]+values[i-1]);
            }else{
                //if W > i-1 we can't chose it
                d2[i][w]=d2[i-1][w];
            }
        }
    }
    //best value it's return d2[n][capacity]
    return d2[n][capacity];
}

const values=[7,200,300];
const weights=[5,20,30];
const capacity=25;
const result = knapsack(values,weights,capacity);
console.log('the best value is :',result);

//Belong to Aryan Alkane

File: backPac.py
def knapsack_01(values, weights, capacity):
    n = len(values)
    dp = [[0] * (capacity + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        for w in range(1, capacity + 1):
            if weights[i - 1] > w:
                dp[i][w] = dp[i - 1][w]
            else:
                dp[i][w] = max(dp[i - 1][w], dp[i - 1][w - weights[i - 1]] + values[i - 1])

    return dp[n][capacity]

# مثال استفاده از تابع
values = [1, 2, 99]
weights = [10, 20, 30]
capacity = 50
result = knapsack_01(values, weights, capacity)
print("   value of pack:", result)

# Belong to aryan Alkane